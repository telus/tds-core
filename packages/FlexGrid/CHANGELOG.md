# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [3.0.11](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.10...@tds/core-flex-grid@3.0.11) (2020-07-22)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.10](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.9...@tds/core-flex-grid@3.0.10) (2020-02-24)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.9](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.8...@tds/core-flex-grid@3.0.9) (2020-02-24)


### Bug Fixes

* **core-flex-grid:** prevent invisible hairline divider w/ mdOffset+ ([9cdec38](https://github.com/telusdigital/tds/commit/9cdec38))





## [3.0.8](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.7...@tds/core-flex-grid@3.0.8) (2020-02-06)


### Bug Fixes

* **core-flex-grid:** prevent incorrect width in box ([653a725](https://github.com/telusdigital/tds/commit/653a725))





## [3.0.7](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.6...@tds/core-flex-grid@3.0.7) (2020-02-03)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.6](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.5...@tds/core-flex-grid@3.0.6) (2020-01-27)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.5](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.4...@tds/core-flex-grid@3.0.5) (2020-01-23)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.4](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.3...@tds/core-flex-grid@3.0.4) (2020-01-22)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.3](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.2...@tds/core-flex-grid@3.0.3) (2019-09-06)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.2](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.1...@tds/core-flex-grid@3.0.2) (2019-08-22)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.0...@tds/core-flex-grid@3.0.1) (2019-08-21)


### Bug Fixes

* **core-flex-grid:** convert px units to rem in flexgrid container ([0da7ffc](https://github.com/telusdigital/tds/commit/0da7ffc))





# [3.0.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.4.1...@tds/core-flex-grid@3.0.0) (2019-07-26)


### Features

* **core-flex-grid:** upgrade to styled-components ([8bf525a](https://github.com/telusdigital/tds/commit/8bf525a))


### BREAKING CHANGES

* **core-flex-grid:** add styled-components as a peerDependency





## [2.4.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.4.0...@tds/core-flex-grid@2.4.1) (2019-07-26)

**Note:** Version bump only for package @tds/core-flex-grid





# [2.4.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.13...@tds/core-flex-grid@2.4.0) (2019-07-25)


### Features

* **core-flex-grid:** upgrade to styled-components ([734f951](https://github.com/telusdigital/tds/commit/734f951))





## [2.3.13](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.12...@tds/core-flex-grid@2.3.13) (2019-06-04)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.12](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.11...@tds/core-flex-grid@2.3.12) (2019-05-13)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.11](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.10...@tds/core-flex-grid@2.3.11) (2019-05-01)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.10](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.9...@tds/core-flex-grid@2.3.10) (2019-04-22)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.9](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.8...@tds/core-flex-grid@2.3.9) (2019-04-04)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.8](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.7...@tds/core-flex-grid@2.3.8) (2019-03-28)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.7](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.6...@tds/core-flex-grid@2.3.7) (2019-03-13)

### Bug Fixes

- **core-flex-grid:** downgrade react-broadcast ([70864e4](https://github.com/telusdigital/tds/commit/70864e4))

## [2.3.6](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.5...@tds/core-flex-grid@2.3.6) (2019-03-11)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.5](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.4...@tds/core-flex-grid@2.3.5) (2019-02-07)

### Bug Fixes

- **core-flex-grid:** adjust css output to be mobile-first ([431cb43](https://github.com/telusdigital/tds/commit/431cb43)), closes [#899](https://github.com/telusdigital/tds/issues/899)

## [2.3.4](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.3...@tds/core-flex-grid@2.3.4) (2019-02-06)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.3](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.2...@tds/core-flex-grid@2.3.3) (2019-01-10)

**Note:** Version bump only for package @tds/core-flex-grid

<a name="2.3.2"></a>

## [2.3.2](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.1...@tds/core-flex-grid@2.3.2) (2018-12-13)

### Bug Fixes

- **core-flex-grid:** apply flex-shrink 0 to FlexGrid.Row ([326c99b](https://github.com/telusdigital/tds/commit/326c99b)), closes [#783](https://github.com/telusdigital/tds/issues/783)

<a name="2.3.1"></a>

## [2.3.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.0...@tds/core-flex-grid@2.3.1) (2018-12-11)

**Note:** Version bump only for package @tds/core-flex-grid

<a name="2.3.0"></a>

# [2.3.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.2.1...@tds/core-flex-grid@2.3.0) (2018-12-10)

### Features

- **core-flex-grid:** add horizontalAlign prop to Col ([694ad0a](https://github.com/telusdigital/tds/commit/694ad0a))

<a name="2.2.1"></a>

## [2.2.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.2.0...@tds/core-flex-grid@2.2.1) (2018-12-06)

**Note:** Version bump only for package @tds/core-flex-grid

<a name="2.2.0"></a>

# [2.2.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.1.1...@tds/core-flex-grid@2.2.0) (2018-08-09)

### Features

- **core-flex-grid:** add hide props to col ([1f4ed86](https://github.com/telusdigital/tds/commit/1f4ed86))
- **core-flex-grid:** hides column starting from breakpoint ([7d6f267](https://github.com/telusdigital/tds/commit/7d6f267))
- **core-flex-grid:** more abstraction in calculateLevels ([faec59d](https://github.com/telusdigital/tds/commit/faec59d))
- **core-flex-grid:** removed hidden flags and enabled 0 for breakpoints ([867599e](https://github.com/telusdigital/tds/commit/867599e))

<a name="2.1.1"></a>

## [2.1.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.1.0...@tds/core-flex-grid@2.1.1) (2018-08-07)

### Bug Fixes

- **core-flex-grid:** add specificity to styles ([ee0d7ff](https://github.com/telusdigital/tds/commit/ee0d7ff))

<a name="2.1.0"></a>

# [2.1.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.0.2...@tds/core-flex-grid@2.1.0) (2018-07-17)

### Features

- **core-flex-grid:** add reversible grid props ([#681](https://github.com/telusdigital/tds/issues/681)) ([9d10ec6](https://github.com/telusdigital/tds/commit/9d10ec6))

<a name="2.0.2"></a>

## [2.0.2](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.0.1...@tds/core-flex-grid@2.0.2) (2018-06-12)

**Note:** Version bump only for package @tds/core-flex-grid

<a name="2.0.1"></a>

## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.0.0...@tds/core-flex-grid@2.0.1) (2018-06-12)

### Bug Fixes

- **core-flex-grid:** force correct spacing through specificity ([da485d4](https://github.com/telusdigital/tds/commit/da485d4))

<a name="2.0.0"></a>

# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@1.2.1...@tds/core-flex-grid@2.0.0) (2018-06-04)

### Features

- **core-flex-grid:** make limitWidth true by default ([9ce0b69](https://github.com/telusdigital/tds/commit/9ce0b69))
- **core-flex-grid:** remove centre prop ([45ebff9](https://github.com/telusdigital/tds/commit/45ebff9))

### BREAKING CHANGES

- **core-flex-grid:** remove the centre prop
- **core-flex-grid:** set limitWidth prop to true by default

<a name="1.2.1"></a>

## [1.2.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@1.2.0...@tds/core-flex-grid@1.2.1) (2018-05-14)

### Bug Fixes

- **core-flex-grid:** remove old responsive sass ([e21a59c](https://github.com/telusdigital/tds/commit/e21a59c))

<a name="1.2.0"></a>

# [1.2.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@1.1.0...@tds/core-flex-grid@1.2.0) (2018-04-11)

### Features

- **core-flex-grid:** add responsive props ([#566](https://github.com/telusdigital/tds/issues/566)) ([10e492c](https://github.com/telusdigital/tds/commit/10e492c)), closes [#465](https://github.com/telusdigital/tds/issues/465)

<a name="1.1.0"></a>

# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@1.0.0...@tds/core-flex-grid@1.1.0) (2018-03-22)

### Features

- **core-flex-grid:** fix gutterless alignment issue and deprecate centre prop ([#526](https://github.com/telusdigital/tds/issues/526)) ([0e8cbeb](https://github.com/telusdigital/tds/commit/0e8cbeb)), closes [#518](https://github.com/telusdigital/tds/issues/518)
