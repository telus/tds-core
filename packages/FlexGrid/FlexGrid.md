**Use the "Open isolated" button above to view this component in full-width mode.**

The `FlexGrid` system is a thin wrapper over [react-flexbox-grid](https://github.com/roylee0704/react-flexbox-grid),
which implements [flexboxgrid.css](http://flexboxgrid.com/). It uses a series of rows and columns to layout and align
content 2-dimensionally.

**If you are new to or unfamiliar with flexbox, read this [CSS Tricks flexbox guide](https://css-tricks.com/snippets/css/a-guide-to-flexbox/)
for background, terminology, guidelines, and code snippets.**

Below is a quick example of how the components fit together. See the [FlexGrid.Row](#row) and [FlexGrid.Col](#col)
components below for a more in-depth look.

```jsx { "props": { "className": "docs_full-width-playground docs_flex-grid-coloring" } }
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col>
      <Box vertical={2}><Text>1/3</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>2/3</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>3/3</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### How it works

* `FlexGrid` is the top-level container. It centers and horizontally pads your siteâ€™s contents
* `FlexGrid.Row`s are wrappers for columns. Columns have horizontal padding (called a gutter) for controlling the space between them
* Place content inside `FlexGrid.Col` components. These should be the immediate children of rows
* Columns without a specified width will automatically layout as equal width columns
* Column widths can be made responsive by using the responsive props, which match the props used in `Responsive`: `xs`, `sm`,
  `md`, `lg` and `xl`
* The responsive column widths are based on minimum width media queries, so they apply to the specified breakpoint and all
  those above it

### Removing the gutter

Remove the 16px gutter surrounding columns by passing `gutter={false}` to the `FlexGrid` container. This removes the horizontal padding
from all children columns.

```jsx { "props": { "className": "docs_full-width-playground docs_flex-grid-coloring" } }
<FlexGrid gutter={false}>
  <FlexGrid.Row>
    <FlexGrid.Col>
      <Box vertical={2}><Text>1/3 no gutter</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>2/3 no gutter</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>3/3 no gutter</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```
