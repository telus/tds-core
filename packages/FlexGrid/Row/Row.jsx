import React from 'react'
import PropTypes from 'prop-types'

import { Row as ReactFlexboxGridRow } from 'react-flexbox-grid'

import styles from './Row.modules.scss'

import safeRest from '../../../shared/utils/safeRest'
import joinClassNames from '../../../shared/utils/joinClassNames'

const Row = ({
  horizontalAlign,
  verticalAlign,
  distribute,
  xsReverse,
  smReverse,
  mdReverse,
  lgReverse,
  xlReverse,
  children,
  ...rest
}) => {
  const getAlignment = () => {
    return {
      start: horizontalAlign === 'start' ? 'xs' : undefined,
      center: horizontalAlign === 'center' ? 'xs' : undefined,
      end: horizontalAlign === 'end' ? 'xs' : undefined,
      top: verticalAlign === 'top' ? 'xs' : undefined,
      middle: verticalAlign === 'middle' ? 'xs' : undefined,
      bottom: verticalAlign === 'bottom' ? 'xs' : undefined,
    }
  }

  const getDistribution = () => {
    return {
      around: distribute === 'around' ? 'xs' : undefined,
      between: distribute === 'between' ? 'xs' : undefined,
    }
  }

  const calculateReverseLevel = () => {
    const levelToggles = [xsReverse, smReverse, mdReverse, lgReverse, xlReverse]
    const enabledLevels = [false, false, false, false, false]

    for (let toggles = 0; toggles < levelToggles.length; toggles += 1) {
      for (let levels = toggles; levels < enabledLevels.length; levels += 1) {
        if (levelToggles[toggles] !== undefined) {
          enabledLevels[levels] = levelToggles[toggles]
        }
      }
    }

    return enabledLevels
  }

  const reverseLevel = calculateReverseLevel()

  return (
    <ReactFlexboxGridRow
      {...safeRest(rest)}
      {...getAlignment()}
      {...getDistribution()}
      className={joinClassNames(
        styles.flexRow,
        reverseLevel[0] ? styles.xsReverse : styles.xsReverseCancel,
        reverseLevel[1] ? styles.smReverse : styles.smReverseCancel,
        reverseLevel[2] ? styles.mdReverse : styles.mdReverseCancel,
        reverseLevel[3] ? styles.lgReverse : styles.lgReverseCancel,
        reverseLevel[4] ? styles.xlReverse : styles.xlReverseCancel
      )}
    >
      {children}
    </ReactFlexboxGridRow>
  )
}

Row.propTypes = {
  /**
   * Align columns horizontally within their row.
   */
  horizontalAlign: PropTypes.oneOf(['start', 'center', 'end']),
  /**
   * Align columns vertically within their row.
   */
  verticalAlign: PropTypes.oneOf(['top', 'middle', 'bottom']),
  /**
   * Distribute empty space around columns.
   */
  distribute: PropTypes.oneOf(['around', 'between']),
  /**
   * Choose if the item order should be reversed within the 'xs' breakpoint range.
   */
  xsReverse: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'sm' breakpoint range.
   */
  smReverse: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'md' breakpoint range.
   */
  mdReverse: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'lg' breakpoint range.
   */
  lgReverse: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'xl' breakpoint range.
   */
  xlReverse: PropTypes.bool,
  children: PropTypes.node.isRequired,
}

Row.defaultProps = {
  horizontalAlign: undefined,
  verticalAlign: undefined,
  distribute: undefined,
  xsReverse: undefined,
  smReverse: undefined,
  mdReverse: undefined,
  lgReverse: undefined,
  xlReverse: undefined,
}

export default Row
