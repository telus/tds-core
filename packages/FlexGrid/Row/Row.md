### Alignment

Use flexbox alignment properties to vertically and horizontally align columns.

```jsx { "props": { "className": "docs_full-width-playground docs_flex-grid-coloring" } }
<FlexGrid>
  <FlexGrid.Row horizontalAlign="start">
    <FlexGrid.Col xs={4} md={2}>
      <Box vertical={2}><Text>Left</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col xs={4} md={2}>
      <Box vertical={2}><Text>aligned</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>

  <FlexGrid.Row horizontalAlign="center">
    <FlexGrid.Col xs={4} md={2}>
      <Box vertical={2}><Text>Center</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col xs={4} md={2}>
      <Box vertical={2}><Text>aligned</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>

  <FlexGrid.Row horizontalAlign="end">
    <FlexGrid.Col xs={4} md={2}>
      <Box vertical={2}><Text>Right</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col xs={4} md={2}>
      <Box vertical={2}><Text>aligned</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

```jsx { "props": { "className": "docs_full-width-playground docs_flex-grid-coloring" } }
<FlexGrid>
  <FlexGrid.Row verticalAlign="top">
    <FlexGrid.Col>
      <Box vertical={2}>
        <Text>Top vertical aligned</Text>
      </Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}>
        <div style={{ height: 100 }} />
      </Box>
    </FlexGrid.Col>
  </FlexGrid.Row>

  <FlexGrid.Row verticalAlign="middle">
    <FlexGrid.Col>
      <Box vertical={2}>
        <Text>Middle vertical aligned</Text>
      </Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}>
        <div style={{ height: 100 }} />
      </Box>
    </FlexGrid.Col>
  </FlexGrid.Row>

  <FlexGrid.Row verticalAlign="bottom">
    <FlexGrid.Col>
      <Box vertical={2}>
        <Text>Bottom vertical aligned</Text>
      </Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}>
        <div style={{ height: 100 }} />
      </Box>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### Distribution

Use flexbox distribution properties to control the negative space around columns.

```jsx { "props": { "className": "docs_full-width-playground docs_flex-grid-coloring" } }
<FlexGrid>
  <FlexGrid.Row distribute="around">
    <FlexGrid.Col xs={3}>
      <Box vertical={2}><Text>Around</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col xs={3}>
      <Box vertical={2}><Text>Around</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col xs={3}>
      <Box vertical={2}><Text>Around</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>

  <FlexGrid.Row distribute="between">
    <FlexGrid.Col xs={3}>
      <Box vertical={2}><Text>Between</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col xs={3}>
      <Box vertical={2}><Text>Between</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col xs={3}>
      <Box vertical={2}><Text>Between</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### Reversing item order

On certain layouts, the order of items within the row may need to be reversed. To do this, set the corresponding prop for the target breakpoint range as `true`. If a breakpoint is defined as `true`, it will affect all other breakpoints after it until a latter breakpoint assigned as `false` is found.

In the following example, the items are set to reverse until the `md` breakpoint. Then, at the `xl` breakpoint, the items will reverse again.

```jsx { "props": { "className": "docs_full-width-playground docs_flex-grid-coloring" } }
<FlexGrid>
  <FlexGrid.Row xsReverse={true} mdReverse={false} xlReverse={true}>
    <FlexGrid.Col>
      <Box vertical={2}><Text>1</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>2</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>3</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```
