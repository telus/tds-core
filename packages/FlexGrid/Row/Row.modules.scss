@import '~@tds/core-responsive/responsive';

div.flexRow {
  margin: 0 auto;
  width: 100%;
}

.xsReverse {
  @include mq($until: sm) {
    flex-direction: row-reverse;
  }
}

.smReverse {
  @include mq($from: sm) {
    flex-direction: row-reverse;
  }
}

.mdReverse {
  @include mq($from: md) {
    flex-direction: row-reverse;
  }
}

.lgReverse {
  @include mq($from: lg) {
    flex-direction: row-reverse;
  }
}

.xlReverse {
  @include mq($from: xl) {
    flex-direction: row-reverse;
  }
}

.xsReverseCancel {
  @include mq($until: sm) {
    flex-direction: row;
  }
}

.smReverseCancel {
  @include mq($from: sm, $until: md) {
    flex-direction: row;
  }
}

.mdReverseCancel {
  @include mq($from: md, $until: lg) {
    flex-direction: row;
  }
}

.lgReverseCancel {
  @include mq($from: lg, $until: xl) {
    flex-direction: row;
  }
}

.xlReverseCancel {
  @include mq($from: xl) {
    flex-direction: row;
  }
}
