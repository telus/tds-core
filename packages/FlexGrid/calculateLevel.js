const calculateLevel = (xs, sm, md, lg, xl) => {
  const levelToggles = [xs, sm, md, lg, xl]
  const enabledLevels = [false, false, false, false, false]

  for (let toggles = 0; toggles < levelToggles.length; toggles += 1) {
    for (let levels = toggles; levels < enabledLevels.length; levels += 1) {
      if (levelToggles[toggles] !== undefined) {
        enabledLevels[levels] = levelToggles[toggles]
      }
    }
  }

  return enabledLevels
}

export default calculateLevel
