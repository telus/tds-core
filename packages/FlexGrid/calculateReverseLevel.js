const calculateReverseLevel = (xsReverse, smReverse, mdReverse, lgReverse, xlReverse) => {
  const levelToggles = [xsReverse, smReverse, mdReverse, lgReverse, xlReverse]
  const enabledLevels = [false, false, false, false, false]

  for (let toggles = 0; toggles < levelToggles.length; toggles += 1) {
    for (let levels = toggles; levels < enabledLevels.length; levels += 1) {
      if (levelToggles[toggles] !== undefined) {
        enabledLevels[levels] = levelToggles[toggles]
      }
    }
  }

  return enabledLevels
}

export default calculateReverseLevel
