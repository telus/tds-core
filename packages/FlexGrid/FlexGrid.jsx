import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast } from 'react-broadcast'
import { Grid } from 'react-flexbox-grid'

import Col from './Col/Col'
import Row from './Row/Row'

import safeRest from '../../shared/utils/safeRest'
import joinClassNames from '../../shared/utils/joinClassNames'

import styles from './FlexGrid.modules.scss'

/**
 * A mobile-first flexbox grid.
 *
 * @version ./package.json
 */
const FlexGrid = ({
  limitWidth,
  gutter,
  xsReverse,
  smReverse,
  mdReverse,
  lgReverse,
  xlReverse,
  children,
  ...rest
}) => {
  const calculateReverseLevel = () => {
    const levelToggles = [xsReverse, smReverse, mdReverse, lgReverse, xlReverse]
    const enabledLevels = [false, false, false, false, false]

    for (let toggles = 0; toggles < levelToggles.length; toggles += 1) {
      for (let levels = toggles; levels < enabledLevels.length; levels += 1) {
        if (levelToggles[toggles] !== undefined) {
          enabledLevels[levels] = levelToggles[toggles]
        }
      }
    }

    return enabledLevels
  }

  const reverseLevel = calculateReverseLevel()

  return (
    <Broadcast channel="flex-grid" value={gutter}>
      <Grid
        {...safeRest(rest)}
        fluid
        className={joinClassNames(
          styles.flexGrid,
          limitWidth && styles.limitWidth,
          reverseLevel[0] ? styles.xsReverse : styles.xsReverseCancel,
          reverseLevel[1] ? styles.smReverse : styles.smReverseCancel,
          reverseLevel[2] ? styles.mdReverse : styles.mdReverseCancel,
          reverseLevel[3] ? styles.lgReverse : styles.lgReverseCancel,
          reverseLevel[4] ? styles.xlReverse : styles.xlReverseCancel
        )}
      >
        {children}
      </Grid>
    </Broadcast>
  )
}

FlexGrid.propTypes = {
  /**
   * Whether or not to give the grid a fixed width. This also centres the grid horizontally.
   */
  limitWidth: PropTypes.bool,
  /**
   * Whether or not to include gutters in between columns.
   */
  gutter: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'xs' breakpoint range.
   */
  xsReverse: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'sm' breakpoint range.
   */
  smReverse: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'md' breakpoint range.
   */
  mdReverse: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'lg' breakpoint range.
   */
  lgReverse: PropTypes.bool,
  /**
   * Choose if the item order should be reversed within the 'xl' breakpoint range.
   */
  xlReverse: PropTypes.bool,
  /**
   * The rows and columns of the Grid. Will typically be `FlexGrid.Row` and `FlexGrid.Col` components.
   */
  children: PropTypes.node.isRequired,
}

FlexGrid.defaultProps = {
  limitWidth: true,
  gutter: true,
  xsReverse: undefined,
  smReverse: undefined,
  mdReverse: undefined,
  lgReverse: undefined,
  xlReverse: undefined,
}

FlexGrid.Row = Row
FlexGrid.Col = Col

export default FlexGrid
