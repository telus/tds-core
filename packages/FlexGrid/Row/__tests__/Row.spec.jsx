import React from 'react'
import { mount } from 'enzyme'
import { Row as ReactFlexboxGridRow } from 'react-flexbox-grid'

import FlexGrid from '../../FlexGrid'
import Row from '../Row'

describe('Row', () => {
  const doMount = (props = {}) => {
    const wrapper = mount(
      <FlexGrid>
        <Row {...props}>Some content</Row>)
      </FlexGrid>
    )
    return wrapper.find(ReactFlexboxGridRow)
  }

  it('renders', () => {
    const row = doMount()
    expect(row).toMatchSnapshot()
  })

  it('renders a react-flexbox-grid Row component', () => {
    const row = doMount()

    expect(row).toMatchSelector(ReactFlexboxGridRow)
  })

  it('passes additional attributes to the react-flexbox-grid Row', () => {
    const row = doMount({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(row).toHaveProp('id', 'the-id')
    expect(row).toHaveProp('data-some-attr', 'some value')
  })

  describe('alignment', () => {
    it('can align columns horizontally to the start, center, or end', () => {
      let row = doMount({ horizontalAlign: 'start' })
      expect(row).toHaveProp('start', 'xs')

      row = doMount({ horizontalAlign: 'center' })
      expect(row).toHaveProp('center', 'xs')

      row = doMount({ horizontalAlign: 'end' })
      expect(row).toHaveProp('end', 'xs')
    })

    it('can align columns vertically to the top, middle, or bottom', () => {
      let row = doMount({ verticalAlign: 'top' })
      expect(row).toHaveProp('top', 'xs')

      row = doMount({ verticalAlign: 'middle' })
      expect(row).toHaveProp('middle', 'xs')

      row = doMount({ verticalAlign: 'bottom' })
      expect(row).toHaveProp('bottom', 'xs')
    })

    it('can be horizontally and vertically aligned', () => {
      const row = doMount({ horizontalAlign: 'start', verticalAlign: 'top' })

      expect(row).toHaveProp('start', 'xs')
      expect(row).toHaveProp('top', 'xs')
    })
  })

  it('can distribute columns around or between', () => {
    let row = doMount({ distribute: 'around' })
    expect(row).toHaveProp('around', 'xs')

    row = doMount({ distribute: 'between' })
    expect(row).toHaveProp('between', 'xs')
  })

  it('supports responsive reversal', () => {
    const row = doMount({
      xsReverse: true,
      smReverse: true,
      mdReverse: true,
      lgReverse: false,
      xlReverse: true,
    })

    expect(row.hasClass('xsReverse')).toEqual(true)
    expect(row.hasClass('smReverse')).toEqual(true)
    expect(row.hasClass('mdReverse')).toEqual(true)
    expect(row.hasClass('lgReverseCancel')).toEqual(true)
    expect(row.hasClass('xlReverse')).toEqual(true)
  })

  it('does not allow custom CSS', () => {
    const row = doMount({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(row).not.toHaveProp('className', 'my-custom-class')
    expect(row).not.toHaveProp('style', { color: 'hotpink' })
  })
})
