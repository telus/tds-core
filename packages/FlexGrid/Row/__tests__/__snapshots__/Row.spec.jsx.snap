// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row renders 1`] = `
<Row
  className="flexRow xsReverseCancel smReverseCancel mdReverseCancel lgReverseCancel xlReverseCancel"
>
  <div
    className="flexRow xsReverseCancel smReverseCancel mdReverseCancel lgReverseCancel xlReverseCancel row"
  >
    Some content
  </div>
</Row>
`;
